### DISCUSSION TITLE
refined support in 0.7

### DISCUSSION THREAD

**Discussion by erikerlandson** [erikerlandson/coulomb#385] [Ideas]
cc @armanbilge  @cquiroz @benhutchinson

I'm preparing to add support for `refined` in 0.6, and I wanted to bounce some thoughts off you.

In 0.5 (scala 2), I supported some fairly rich logic around manipulating refined predicates. Like adding ">= 1" and ">= 1" could yield an output with ">= 2", etc.

While I have some interest in continuing to support that kind of clever type manipulation, I am wondering what the use cases really are.

Because the reality is that I can support the following with much less effort:

- positive
- non-negative
- negative
- non-positive

Those four I can support with just a few additional algebraic rules for `AdditiveSemigroup` (and friends) and  `ValueResolution` , and I have a suspicion that these cover most use cases.

I wanted to see what you think of that as an initial offering.  It doesn't preclude support for more advanced refined predicate logic in the future, but I could get this going pretty fast, and the fancy stuff will be significantly more work.

**Comment by benhutchison** [comment]
Agree that the 4 you mention are key and more relevant than other
predicates because they rest on mathematical meaning, and not just
arbitrary boundary points.

There are two others I think are worthy of consideration:
- Fractions [0, 1] Some quite meaningful properties. See also
https://youtu.be/lLUazJS-65k?t=670
- Fractional positive numbers >= 1. A sort of complement to fractions, when
viewed from the perspective of multiplication. Ie multiplying by such a
number grows the output, while multiplying by a fraction shrinks it.

**Reply by erikerlandson** [reply]
yes, those should also behave as groups or semigroups for multiplication

**Comment by erikerlandson** [comment]
https://github.com/erikerlandson/coulomb/pull/392

**Comment by benhutchison** [comment]
🤔💭.. a foundation for a future magnitude operation defined over Quantity,
that yields a positive amount associated with some units..

**Comment by erikerlandson** [comment]
Still thinking about what an implementation of the more general logic might look like. Taking addition as an example, an extension to refined might look like the following

```scala
    transparent inline given ctx_add_refined_2V2U[VL, UL, VR, UR, PL, PR](using
        vres: ValueResolution[VL, VR],
        icl: Conversion[Quantity[VL, UL], Quantity[vres.VO, UL]],
        icr: Conversion[Quantity[VR, UR], Quantity[vres.VO, UL]],
        alg: AdditiveSemigroup[vres.VO],
        pred: ResolvePredicateAdd[vres.VO, PL, PR]
    ): Add[Refined[VL, PL], UL, Refined[VR, PR], UR] =
        new infra.AddNC((ql: Quantity[Refined[VL, PL], UL], qr: Quantity[Refined[VR, PR], UR]) =>
            refinedV[pred.PO].unsafeFrom(alg.plus(icl(ql).value, icr(qr).value)).withUnit[UL]
        )
```

where `ResolvePredicateAdd[PL, PR]` can figure out such logic as `">= 2" + ">= 3" => ">= 5"`, etc, and will degenerate to `True` for any situations where it can draw no conclusions.  Here, I think the new scala3 metaprogramming may make this nice to implement.

There are also the cases of mixing refined with unrefined values, where the output type should also then be unrefined, or possibly `Refined[V, True]`

```
refined + refined => refined  (may degenerate to refined[V, true])
refined + unrefined => unrefined
unrefined + refined => unrefined
```

**Reply by erikerlandson** [reply]
generalized rules like this would probably be incompatible with the more restrictive algebra based rules above, since there will be cases where either could match and the compiler will fail with ambiguous implicit resolutions.  So a user would have to import one or the other.

**Reply by erikerlandson** [reply]
Almost forgot, conversion factors also impact these.  If I have `Quantity[Refined[V, >= x], Meter]` then if I convert that to units of `Kilo*Meter` I now have `Quantity[Refined[V, >= x/1000], Kilo*Meter]`

So the `Conversion` steps above would need rules here, it's not sufficient to convert in "unrefined" space

**Comment by armanbilge** [comment]
Sorry, took me a little while to circle back to this :) thank you so much for opening the discussion!

Since we need refined+coulomb integration, but it wasn't available during the milestone series, we added some of our own helpers here:

https://github.com/gemini-hlsw/lucuma-core/blob/6a08df584dc84b1838c919ad7202b5213b04fe59/modules/core/shared/src/main/scala/lucuma/core/math/units.scala#L139-L158

**Reply by erikerlandson** [reply]
I am planning to include these in something like `coulomb.refined` or `coulomb.refined.syntax`

**Reply by erikerlandson** [reply]
@armanbilge  how is refineMV working with scala 3?
https://github.com/gemini-hlsw/lucuma-core/blob/6a08df584dc84b1838c919ad7202b5213b04fe59/modules/core/shared/src/main/scala/lucuma/core/math/units.scala#L150

this is a known issue upstream: https://github.com/fthomas/refined/issues/932

**Reply by armanbilge** [reply]
We wrote our own implementation. I linked it in that thread.
https://github.com/gemini-hlsw/lucuma-refined/blob/fa0b3d9a990e1ae94f56f1423473d95e6fcdc5f5/refined/src/main/scala/main/lucuma/refined.scala#L26-L27

**Reply by erikerlandson** [reply]
If I understand what you're doing here, you are proposing an upstream replacement of `Validate` with `Predicate`, as you implemented it in your library?

Given that the state of compile-time `refineMV` in scala 3 is a bit up in the air, I'm feeling inclined to wait until it's available upstream before I write any `coulomb.refined.syntax` around it. I saw you also had a `refineV` style that returns `Quantity[Either[String, Refined[V, P]], U]`, and I could provide that if you're interested, and even do some algebras with it (see discussion below).

**Comment by benhutchison** [comment]
While arbitrary refined predicates like `>= 2` are cool, I'm personally unconvinced how useful they are in practice. 

I think the refined types that will prove to have enduring value will be where the refinement reflects an underlying mathematical structure.

Probably half the value of the refined type universe is probably embodied in a single predicate, non-negativity. IMO, naturally non-negative quantities almost always arise because we're modeling a magnitude or amount of something.

I'd advocate for a slightly more "opinionated" approach, supporting specific refined sets that feature closed operations defined over them.

Whereas, for a set like `>= 2 && < 3`, I can't think of a single operation on such a value that preserves the refinement. I guess it's perhaps useful from a types-as-documentation standpoint, and perhaps to use metaprogramming to derive validators...?

**Reply by erikerlandson** [reply]
Agree, as you can see from my thought experiments, supporting "non-algebraic" predicates is a step-function increase in complexity, for what I suspect is very limited benefit.  Its primary appeal to me is the "type system cool factor" but I'm going to resist, especially for a first "mvp" release.

**Reply by erikerlandson** [reply]
another nice property of `Positive` and `NonNegative` is that `refined` treats these the same for all raw numeric types, so for example when I do value conversions, I do not have to change `Greater[0]` to `Greater[0.0]` - which is possible but I'd  have to do it in the meta programming, whereas in these cases I can define relatively simple implicit rules.

**Comment by erikerlandson** [comment]
one thing I *think* I could do pretty cleanly is support `Either[L, Refined[V, P]]` as an additional overlay.

to wit, the output of `refineV` is an `Either`:
```scala
scala> refineV[Positive](1)
val res0: Either[String, Int Refined eu.timepit.refined.numeric.Positive] = Right(1)
```

It's possible to define and `AdditiveSemigroup` on this `Either` type.
```
Left + Left => Left
Left + Right(y) => Left
Right(x) + Left => Left
Right(x) + Right(y) => Right(x+y), where x+y is Refined[V, P] + Refined[V, P] - the algebras I am currently defining.
```

**Reply by erikerlandson** [reply]
I'm going to get the `Refined[V, P]` layer finished and tested, but once that is working I'd be in a position to experiment with this `Either` layer

**Reply by erikerlandson** [reply]
So in this algebra our quantity objects look like:
```scala
Quantity[Either[String, Refined[V, P]], U]
```
:joy:

**Reply by erikerlandson** [reply]
The power of algebras - they compose very easily :tada:

**Reply by erikerlandson** [reply]
I almost wonder if there is some `cats` rule already defined here?  Like, for any type `Either[L, R]`, as long as there is some `AdditiveSemigroup[L]` and `AdditiveSemigroup[R]`, the rule is quite general

```scala
given given_AS_Either[L, R](using
    asL: AdditiveSemigroup[L],
    asR: AdditiveSemigroup[R]
): AdditiveSemigroup[Either[L, R]] =
    new AdditiveSemigroup[Either[L, R]]:
        def plus(x: Either[L, R], y: Either[L, R]): Either[L, R] =
            (x, y) match
                case (Left(xl), Left(yl)) => Left(asL.plus(xl, yl))
                case (Left(xl), Right(_)) => Left(xl)
                case (Right(_), Left(yl)) => Left(yl)
                case (Right(xr), Right(yr)) => Right(asR.plus(xr, yr))
```

**Comment by benhutchison** [comment]
For me, the aspect of refinement support that is most exciting is the
fairly rich set of operations that naturally yield or preserve positive
outputs.

Addition, multiplication and division of two positive numbers yields a
positive output. Exponentiation of a positive with any number is positive.
A form of subtraction that yields the distance between two numbers
(unsigned) is always positive.

There will be cases at the boundary of programs to work with Either, but
within the domain model, I hope to establish positivity and then keep it
that way... :)

But I guess this ability to auto-lift Positive => Positive operations into
Either => Either operations makes it close to zero cost to hook up a
refined domain model to an unrefined outside world, where eg strings are
being parsed off the wire..

**Reply by erikerlandson** [reply]
good point, you reminded me that positive numbers are a multiplicative group, so I can support division there

**Reply by erikerlandson** [reply]
> A form of subtraction that yields the distance between two numbers
(unsigned) is always positive.

So, distance metric in 1D?

**Reply by benhutchison** [reply]
Either a distance metric (my doubt is that I dont recall there being such a thing in Cats Algebra ?), or else an absolute value combined with subtraction might be simpler way of getting to something guaranteed to be positive, ie |a - b|
